src/controller/CommissioningWindowOpener.h

```diff
@@ -0,0 +1,172 @@
+
+class CommissioningWindowOpener
+{
+public:
+    CommissioningWindowOpener(DeviceController * controller) :
+        mController(controller), mDeviceConnected(&OnDeviceConnectedCallback, this),
+        mDeviceConnectionFailure(&OnDeviceConnectionFailureCallback, this)
+    {}
+
+    enum class CommissioningWindowOption : uint8_t
+    {
+        kOriginalSetupCode = 0,
+        kTokenWithRandomPIN,
+        kTokenWithProvidedPIN,
+    };
+
+    /*
+     * @brief
+     *   Try to look up the device attached to our controller with the given
+     *   node id and ask it to re-enter commissioning mode with its original
+     *   PASE verifier, discriminator, etc. The device will exit commissioning
+     *   mode after a successful commissioning, or after the given `timeout`
+     *   time.
+     *
+     * @param[in] deviceId The device Id.
+     * @param[in] timeout  The commissioning mode should terminate after this much time.
+     * @param[in] callback The callback to call once the commissioning window is
+     *                     open or if an error occurs.
+     */
+    CHIP_ERROR OpenBasicCommissioningWindow(NodeId deviceId, uint16_t timeout,
+                                            Callback::Callback<OnOpenBasicCommissioningWindow> * callback);
+
+    /**
+     * @brief
+     *   Try to look up the device attached to our controller with the given
+     *   node id and ask it to re-enter commissioning mode with a PASE verifier
+     *   derived from the given information and the given discriminator. The
+     *   device will exit commissioning mode after a successful commissioning,
+     *   or after the given `timeout` time.
+     *
+     * @param[in] deviceId      The device Id.
+     * @param[in] timeout       The commissioning mode should terminate after this much time.
+     * @param[in] iteration     The PAKE iteration count associated with the PAKE Passcode ID and ephemeral
+     *                          PAKE passcode verifier to be used for this commissioning.
+     * @param[in] discriminator The long discriminator for the DNS-SD advertisement.
+     * @param[in] setupPIN      The setup PIN to use, or NullOptional to use a randomly-generated one.
+     * @param[in] callback      The function to be called on success or failure of opening of commissioning window.
+     * @param[out] payload      The setup payload, not including the VID/PID bits, that is generated based on the
+     *                          passed-in information.
+     *
+     * @param[in] readVIDPIDAttributes Should the API internally read VID and PID from the device while opening the
+     *                                 commissioning window.  If this argument is `true`, the API will read VID and
+     *                                 PID from the device and include them in the setup payload passed to the
+     *                                 callback.
+     *
+     * @return CHIP_ERROR         CHIP_NO_ERROR on success, or corresponding error
```

Comment:
`@return` doccomment does not help in reading code: it is obvious that returning CHIP_ERROR returns NO_ERROR on success and an error on failure. Is there any value in keeping such things?